src/rc_string.c

```diff
@@ -7,4 +7,168 @@
+
+MU_DEFINE_ENUM(STRING_STORAGE_TYPE, STRING_STORAGE_TYPE_VALUES)
+
+// In order not to expose the storage type to the user (as it is something internal to the implementation),
+// we're going to have another structure that wraps the RC_STRING and the storage type
+// It is imperative that the first member is the RC_STRING structure though, so it matches the strcuture that
+// we return to the user
+typedef struct RC_STRING_INTERNAL_TAG
+{
+    RC_STRING rc_string;
+    STRING_STORAGE_TYPE storage_type;
+    RC_STRING_FREE_FUNC free_func;
+    void* free_func_context;
+} RC_STRING_INTERNAL;
+
+#define RC_STRING_INTERNAL_FROM_RC_STRING(rc_string_content) \
+    (RC_STRING_INTERNAL*)(void*)(rc_string_content);
+
+static void rc_string_dispose(RC_STRING* content)
+{
+    RC_STRING_INTERNAL* rc_string_internal = ((void*)content);
+    switch (rc_string_internal->storage_type)
+    {
+    default:
+        LogError("String created by Chuck Norris");
+        break;
+
+    case STRING_STORAGE_TYPE_COPIED:
+        break;
+
+    case STRING_STORAGE_TYPE_MOVED:
+        free((void*)rc_string_internal->rc_string.string);
+        break;
+
+    case STRING_STORAGE_TYPE_WITH_CUSTOM_FREE:
+        /* Codes_SRS_RC_STRING_01_018: [ When the THANDLE(RC_STRING) reference count reaches 0, free_func shall be called with free_func_context to free the memory used by string. ]*/
+        rc_string_internal->free_func(rc_string_internal->free_func_context);
+        break;
+    }
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(RC_STRING), rc_string_create, const char*, string)
+{
+    THANDLE(RC_STRING) result = NULL;
+
+    if (string == NULL)
+    {
+        /*Codes_SRS_RC_STRING_01_001: [ If string is NULL, rc_string_create shall fail and return NULL. ]*/
+        LogError("Invalid arguments: const char*, string=%s", MU_P_OR_NULL(string));
+    }
+    else
+    {
+        /* Codes_SRS_RC_STRING_01_002: [ Otherwise, rc_string_create shall determine the length of string. ]*/
+        size_t string_length = strlen(string);
+        size_t string_length_with_terminator = string_length + 1;
+
+        /* Codes_SRS_RC_STRING_01_003: [ rc_string_create shall allocate memory for the THANDLE(RC_STRING), ensuring all the bytes in string can be copied (including the zero terminator). ]*/
+        THANDLE(RC_STRING) temp_result = THANDLE_MALLOC_WITH_EXTRA_SIZE(RC_STRING)(rc_string_dispose, sizeof(RC_STRING_INTERNAL) - sizeof(RC_STRING) + string_length_with_terminator);
+        if (temp_result == NULL)
+        {
+            /* Codes_SRS_RC_STRING_01_006: [ If any error occurs, rc_string_create shall fail and return NULL. ]*/
+            LogError("THANDLE_MALLOC_WITH_EXTRA_SIZE(RC_STRING) failed");
+        }
+        else
+        {
+            RC_STRING_INTERNAL* rc_string_internal = RC_STRING_INTERNAL_FROM_RC_STRING(THANDLE_GET_T(RC_STRING)(temp_result));
+            rc_string_internal->rc_string.string = (const char*)(rc_string_internal + 1);
```

Comment:
>rc_string_internal + 1 [](start = 65, length = 22)

would feel a whole lot better if RC_STRING_INTERNAL would have a char content[] flexible array at the end of the structure. We're anyway allocating it like a flexible array so then it might just have it. And it is easier to find "content" among all the pointers.

Reply:
Added

---
In reply to: [460257508](https://github.com/Azure/azure-c-util/pull/52#discussion_r460257508) [](ancestors = 460257508)